test/test_euler.py

```diff
@@ -0,0 +1,472 @@
+from mirgecom.initializers import Vortex2D
+from mirgecom.initializers import Lump
+from mirgecom.boundary import BoundaryBoss
+from meshmode.discretization import Discretization
+from grudge.eager import EagerDGDiscretization
+from pyopencl.tools import (  # noqa
+    pytest_generate_tests_for_pyopencl as pytest_generate_tests,
+)
+
+# Tests go here
+
+
+def test_inviscid_flux():
+
+    ctx = cl.create_some_context(interactive=False)
+    queue = cl.CommandQueue(ctx)
+
+    def scalevec(scalar, vec):
+        # workaround for object array behavior
+        return make_obj_array([ni * scalar for ni in vec])
+
+    cl_ctx = cl.create_some_context()
+    queue = cl.CommandQueue(cl_ctx)
+
+    dim = 2
+    nel_1d = 16
+
+    from meshmode.mesh.generation import generate_regular_rect_mesh
+
+    mesh = generate_regular_rect_mesh(
+        a=(-0.5,) * dim, b=(0.5,) * dim, n=(nel_1d,) * dim
+    )
+
+    order = 3
+
+    if dim == 2:
+        # no deep meaning here, just a fudge factor
+        dt = 0.75 / (nel_1d * order ** 2)
+    elif dim == 3:
+        # no deep meaning here, just a fudge factor
+        dt = 0.45 / (nel_1d * order ** 2)
+    else:
+        raise ValueError("don't have a stable time step guesstimate")
+
+    print("%d elements" % mesh.nelements)
+
+    discr = EagerDGDiscretization(cl_ctx, mesh, order=order)
+
+    rho = cl.clrandom.rand(queue, (mesh.nelements,), dtype=np.float64)
+    rhoE = cl.clrandom.rand(queue, (mesh.nelements,), dtype=np.float64)
+    rhoV = make_obj_array(
+        [
+            cl.clrandom.rand(queue, (mesh.nelements,), dtype=np.float64)
+            for i in range(dim)
+        ]
+    )
+    scal1 = cl.clrandom.rand(queue, (mesh.nelements,), dtype=np.float64)
+    p = cl.clrandom.rand(queue, (mesh.nelements,), dtype=np.float64)
+    ke = 0.5 * np.dot(rhoV, rhoV) / rho
+
+    # ideal single spec
+    gamma = 1.4
+    p = (gamma - 1.0) * (rhoE - ke)
+    escale = (rhoE + p) / rho
+    expected_mass_flux = rhoV
+    expected_energy_flux = scalevec(escale, rhoV)
+    expected_mom_flux = make_obj_array(
+        [
+            (rhoV[i] * rhoV[j] / rho + (p if i == j else 0))
+            for i in range(dim)
+            for j in range(dim)
+        ]
+    )
+
+    q = join_fields(rho, rhoE, rhoV)
+
+    from mirgecom.euler import _inviscid_flux
+
+    flux = _inviscid_flux(discr, q)
+
+    rhoflux = flux[0:dim]
+    rhoEflux = flux[dim : 2 * dim]
+    momflux = flux[2 * dim :]
+
+    # these should be exact, right?
+    for i in range(dim):
+        assert (
+            la.norm(rhoflux[i].get() - expected_mass_flux[i].get())
+            == 0.0
+        )
+        assert (
+            la.norm(rhoEflux[i].get() - expected_energy_flux[i].get())
+            == 0.0
+        )
+    for i in range(dim * dim):
+        assert (
+            la.norm(momflux[i].get() - expected_mom_flux[i].get())
+            == 0.0
+        )
+
+
+def test_facial_flux():
+    cl_ctx = cl.create_some_context()
```

Comment:
`ctx_factory`

Reply:
I keep forgetting to talk with you about the ctx_factory stuff.  When I add use (by uncommenting the ctx_factory stuff in test_euler.py), I get these errors:

> (dgfem) Michaels-MBP:test mtcampbe$ pytest -k vortex test_euler.py 
> ============================================= test session starts ==============================================
> platform darwin -- Python 3.8.3, pytest-5.4.3, py-1.8.2, pluggy-0.13.1
> rootdir: /Users/mtcampbe/CEESD/devel/emirge/mirgecom
> collected 0 items / 1 error                                                                                    
> 
> ==================================================== ERRORS ====================================================
> _____________________________________ ERROR collecting test/test_euler.py ______________________________________
> ../../../../../miniforge3/envs/dgfem/lib/python3.8/site-packages/_pytest/python.py:511: in _importtestmodule
>     mod = self.fspath.pyimport(ensuresyspath=importmode)
> ../../../../../miniforge3/envs/dgfem/lib/python3.8/site-packages/py/_path/local.py:704: in pyimport
>     __import__(modname)
> <frozen importlib._bootstrap>:991: in _find_and_load
>     ???
> <frozen importlib._bootstrap>:975: in _find_and_load_unlocked
>     ???
> <frozen importlib._bootstrap>:671: in _load_unlocked
>     ???
> ../../../../../miniforge3/envs/dgfem/lib/python3.8/site-packages/_pytest/assertion/rewrite.py:143: in exec_module
>     source_stat, co = _rewrite_test(fn, self.config)
> ../../../../../miniforge3/envs/dgfem/lib/python3.8/site-packages/_pytest/assertion/rewrite.py:328: in _rewrite_test
>     tree = ast.parse(source, filename=fn)
> ../../../../../miniforge3/envs/dgfem/lib/python3.8/ast.py:47: in parse
>     return compile(source, filename, mode, flags,
> E     File "/Users/mtcampbe/CEESD/devel/emirge/mirgecom/test/test_euler.py", line 804
> E       PYOPENCL_TEST=port python -m pudb test_euler.py 'test_isentropic_vortex(cl._csc)'
> E                          ^
> E   SyntaxError: invalid syntax
> =========================================== short test summary info ============================================
> ERROR test_euler.py
> !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
> =============================================== 1 error in 0.27s ===============================================
> 

I kind of need this to work (pytest test_euler.py) - to get TEESD to work as-is.  Obviously, we can tweak TEESD to do the right thing if we need to call these tests differently.

Reply:
The error message makes it look like there is shell source code in a Python file:
```
PYOPENCL_TEST=port python -m pudb test_euler.py 'test_isentropic_vortex(cl._csc)'
```
I can see why it would be upset about that.

Reply:
Ah - thanks - it must have happened during cut-and-paste at some point. Here's the error that is currently produced when using the ctx_factory.

> ================================================= FAILURES =================================================
> ______________ test_vortex_rhs[<context factory for <pyopencl.Device 'Iris Pro' on 'Apple'>] _______________
> 
> obj = <grudge.eager.EagerDGDiscretization object at 0x1266ff4f0>, args = (), kwargs = {}, key = ()
> 
>     def wrapper(obj, *args, **kwargs):
>         if kwargs:
>             key = (_HasKwargs, frozenset(six.iteritems(kwargs))) + args
>         else:
>             key = args
>     
>         try:
> >           return getattr(obj, cache_dict_name)[key]
> E           AttributeError: 'EagerDGDiscretization' object has no attribute '_memoize_dic_opposite_face_connection'
> 
> ../../pytools/pytools/__init__.py:599: AttributeError
> 
> During handling of the above exception, another exception occurred:
> 
> ctx_factory = <pyopencl.tools.pytest_generate_tests_for_pyopencl.<locals>.ContextFactory object at 0x114c95970>
> 
> ( .... )
>

> test_euler.py:630: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> ../mirgecom/euler.py:186: in inviscid_operator
>     discr, w_tpair=_interior_trace_pair(discr, w), eos=eos
> ../mirgecom/euler.py:61: in _interior_trace_pair
>     e = with_object_array_or_scalar(
> ../../pytools/pytools/obj_array.py:364: in with_object_array_or_scalar
>     result[i] = f(field[i])
> ../mirgecom/euler.py:62: in <lambda>
>     lambda el: discr.opposite_face_connection()(el.queue, el), i
> ../../pytools/pytools/__init__.py:601: in wrapper
>     result = function(obj, *args, **kwargs)
> ../../grudge/grudge/discretization.py:315: in opposite_face_connection
>     return make_opposite_face_connection(self._interior_faces_connection())
> ../../meshmode/meshmode/discretization/connection/opposite_face.py:422: in make_opposite_face_connection
>     groups[i_tgt_grp].extend(batches)
> ../../meshmode/meshmode/discretization/connection/opposite_face.py:168: in _make_cross_face_batches
>     resid = apply_map(src_unit_nodes) - tgt_bdry_nodes
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 
> unit_nodes = array([[[            nan,             nan],
>         [            nan,             nan],
>         [            nan,     ...       nan,             nan],
>         [            nan,             nan],
>         [            nan,             nan]]])
> 
>     def apply_map(unit_nodes):
>         # unit_nodes: (dim, nelements, ntgt_unit_nodes)
>     
>         # basis_at_unit_nodes
>         basis_at_unit_nodes = np.empty((nsrc_funcs, nelements, ntgt_unit_nodes))
>     
>         for i, f in enumerate(src_grp.basis()):
>             basis_at_unit_nodes[i] = (
>                     f(unit_nodes.reshape(dim, -1))
>                     .reshape(nelements, ntgt_unit_nodes))
>     
>         intp_coeffs = np.einsum("fj,jet->fet", inv_t_vdm, basis_at_unit_nodes)
>     
>         # If we're interpolating 1, we had better get 1 back.
>         one_deviation = np.abs(np.sum(intp_coeffs, axis=0) - 1)
>        assert (one_deviation < tol).all(), np.max(one_deviation)
> E       AssertionError: nan
> 
> ../../meshmode/meshmode/discretization/connection/opposite_face.py:102: AssertionError
> ============================================= warnings summary =============================================
>